Codegen bug with lock - no member "abi"

Reopen #12855 under my name to avoid spamming the original poster as Github keeps notifying an issue author even if they unsubscribe.


> Bad C code is generated with the following Nim code:
> ### Example
> 
> ```nim
> import locks
> 
> var l: Lock
> echo l
> ```
> 
> ### Current Output
> 
> ```
> CC: ltest.nim
> Error: execution of an external compiler program 'gcc -c  -w  -I/home/tay/.choosenim/toolchains/nim-1.0.4/lib -I/home/tay -o /home/tay/.cache/nim/ltest_d/stdlib_dollars.nim.c.o /home/tay/.cache/nim/ltest_d/stdlib_dollars.nim.c' failed with exit code: 1
> 
> /home/tay/.cache/nim/ltest_d/stdlib_dollars.nim.c: In function ‘dollar___3Dq1hcGH7a2iNlnh4hnPbg’:
> /home/tay/.cache/nim/ltest_d/stdlib_dollars.nim.c:219:50: error: ‘pthread_mutex_t’ {aka ‘union <anonymous>’} has no member named ‘abi’
>   219 |  addQuoted__N56rmvBL9a3xqnKq09cKdb3A((&result), x.abi);
>       |                                                  ^
> ```
> 
> ### Expected Output
> 
> Compiler either compiles properly or errors before C code gen
> ### Possible Solution
> 
>     * It looks like there might be an issue with the implementation of `$` for the Lock type
> 
> 
> ### Additional Information
> 
>     * Fedora 31: `Linux localhost.localdomain 5.3.11-300.fc31.x86_64 #1 SMP Tue Nov 12 19:08:07 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
> 
> 
> ```
> $ nim -v
> Nim Compiler Version 1.0.4 [Linux: amd64]
> Compiled at 2019-11-27
> Copyright (c) 2006-2019 by Andreas Rumpf
> 
> git hash: c8998c498f5e2a0874846eb31309e1d1630faca6
> active boot switches: -d:release
> ```

I've encountered a similar error but without using echo. I'll put it here although, it may belong in a separate issue.

It seems that a class that contains both a Lock and a string property cannot be instantiated outside of a proc or similar:

### Example

```nim
import Locks

type
  Test = object
    path: string # Removing this makes both cases work.
    lock: Lock

# A: This is not fine.
var a = Test()

proc main(): void =
  # B: This is fine.
  var b = Test()

main()

```

### Current Output

```
/home/tdely/.cache/nim/main_d/@mmain.nim.c: In function 'NimMainModule':
/home/tdely/.cache/nim/main_d/@mmain.nim.c:217:33: error: 'pthread_mutex_t' {aka 'union <anonymous>'} has no member named 'abi'
  a__9cWLbw5tGuQDnREm7o4KXdA.lock.abi[T1_] = 0;
                                 ^
Error: execution of an external program failed: 'gcc -c  -w -fmax-errors=3   -I/usr/local/lib/nim -I/home/tdely/test -o /home/tdely/.cache/nim/main_d/@mmain.nim.c.o /home/tdely/.cache/nim/main_d/@mmain.nim.c'

```

### Expected Output

I expected it work really.

### Additional Information

```
Nim Compiler Version 1.4.2 [Linux: amd64]
Compiled at 2020-11-30
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: 3fb5157ab1b666a5a5c34efde0f357a82d433d04
active boot switches: -d:release
```

--- a/lib/core/locks.nim
+++ b/lib/core/locks.nim
@@ -28,11 +28,6 @@ type
 
 {.push stackTrace: off.}
 
-
-proc `$`*(lock: Lock): string =
-  # workaround bug #14873
-  result = "()"
-
 proc initLock*(lock: var Lock) {.inline.} =
   ## Initializes the given lock.
   when not defined(js):

--- a/lib/system/syslocks.nim
+++ b/lib/system/syslocks.nim
@@ -102,26 +102,18 @@ else:
     SysLockObj {.importc: "pthread_mutex_t", pure, final,
                header: """#include <sys/types.h>
                           #include <pthread.h>""".} = object
-      when defined(linux) and defined(amd64):
-        abi: array[40 div sizeof(clong), clong]
 
     SysLockAttr {.importc: "pthread_mutexattr_t", pure, final
                header: """#include <sys/types.h>
                           #include <pthread.h>""".} = object
-      when defined(linux) and defined(amd64):
-        abi: array[4 div sizeof(cint), cint]  # actually a cint
 
     SysCondObj {.importc: "pthread_cond_t", pure, final,
                header: """#include <sys/types.h>
                           #include <pthread.h>""".} = object
-      when defined(linux) and defined(amd64):
-        abi: array[48 div sizeof(clonglong), clonglong]
 
     SysCondAttr {.importc: "pthread_condattr_t", pure, final
                header: """#include <sys/types.h>
                           #include <pthread.h>""".} = object
-      when defined(linux) and defined(amd64):
-        abi: array[4 div sizeof(cint), cint]  # actually a cint
 
     SysLockType = distinct cint
 

--- a/tests/stdlib/uselocks.nim
+++ b/tests/stdlib/uselocks.nim
@@ -11,5 +11,6 @@ proc use* (m: var MyType): int =
     result = 3
 
 block:
+  # bug #14873
   var l: Lock
   doAssert $l == "()"

--- a/tests/stdlib/uselocks.nim
+++ b/tests/stdlib/uselocks.nim
@@ -14,3 +14,16 @@ block:
   # bug #14873
   var l: Lock
   doAssert $l == "()"
+
+when true: # intentional
+  # bug https://github.com/nim-lang/Nim/issues/14873#issuecomment-784241605
+  type
+    Test = object
+      path: string # Removing this makes both cases work.
+      lock: Lock
+  # A: This is not fine.
+  var a = Test()
+  proc main(): void =
+    # B: This is fine.
+    var b = Test()
+  main()

--- a/tests/stdlib/uselocks.nim
+++ b/tests/stdlib/uselocks.nim
@@ -13,7 +13,9 @@ proc use* (m: var MyType): int =
 block:
   # bug #14873
   var l: Lock
-  doAssert $l == "()"
+  doAssert ($l).len > 0
+    # on posix, "()", on windows, something else, but that shouldn't be part of the spec
+    # what matters is that `$` doesn't cause the codegen bug mentioned
 
 when true: # intentional
   # bug https://github.com/nim-lang/Nim/issues/14873#issuecomment-784241605

